[console] 6.0 API and ace refactor

bmcconaghy opened the pull request:
This refactor has two primary goals:
--update Elasticsearch API support to 6.0 level for autocomplete
--remove webpackShims files for ace and associated libraries

Closes https://github.com/elastic/kibana/issues/6879 https://github.com/elastic/kibana/issues/15144 https://github.com/elastic/kibana/issues/7883 https://github.com/elastic/kibana/issues/12108 https://github.com/elastic/kibana/issues/15073 https://github.com/elastic/kibana/issues/8791 https://github.com/elastic/kibana/issues/10264

spalger commented:
Personal opinion, I'd like if the two primary goals of this pr were two individual pull requests.

spalger commented on src/core_plugins/console/api_server/api.js:
```diff
@@ -1,45 +1,52 @@
 
-    var copiedDescription = {};
-    _.extend(copiedDescription, description || {});
+    const url_params_def = {};
```
Please use camelCase here

spalger commented on src/core_plugins/console/api_server/api.js:
```diff
@@ -1,45 +1,52 @@
-      "endpoints": this.endpoints
-    }
-  };
+      'name': this.name,
```
No need for quotes on these keys.

spalger commented on src/core_plugins/console/api_server/es_6_0.js:
```diff
@@ -0,0 +1,27 @@
+Object.keys(spec).forEach(endpoint => {
+  ES_6_0.addEndpointDescription(endpoint, spec[endpoint]);
+});
+_.each(parts, function (apiSection) {
```
Is there a reason we don't just `require('./es_6_0/aliases')(ES_6_0)` rather than require into an array first then iterate and call later?

If we can't get rid of the separation, mind swapping out `_.each` with `[].forEach()`?

bmcconaghy replied:
Legacy of old code style.  We can fix in a separate PR to address all of that kind of stuff in one go.

spalger commented on src/core_plugins/console/api_server/es_6_0/aggregations.js:
```diff
@@ -0,0 +1,514 @@
+/*eslint camelcase: 0*/
```
What's the value in disabling the camelcase rule for one file? Why not just use camelCase?

bmcconaghy replied:
the API uses camel case.

spalger replied:
This override is just for local variable names which don't need to be snake_case, even if the API does expect snake_case.

spalger commented on src/core_plugins/console/api_server/es_6_0/query.js:
```diff
@@ -25,47 +25,47 @@ var SPAN_QUERIES = {
-    "boost_factor": 2.0,
-    "random_score": {
-      "seed": 314159265359
+  'boost_factor': 2.0,
```
Perhaps we should add the `quote-props: ['error', 'as-needed']` to our eslint config, but these keys don't need them.

Thanks for switching things to single quotes btw.

spalger commented on src/core_plugins/console/api_server/es_6_0/query.js:
```diff
@@ -586,33 +586,33 @@ export default function (api) {
       },
-      SCORING_FUNCS
+      query: {},
+      functions: [
```
Looks like in a couple places, like here, things are being pulled out of the `__template` section. What's the reasoning behind that?

spalger commented on src/core_plugins/console/api_server/es_6_0/reindex.js:
```diff
@@ -1,18 +1,10 @@
     patterns: [
       '_reindex'
     ],
-    url_params: {
```
The reindex api still [seems to still support these parameters](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/reindex.json#L10-L37), why remove them?

spalger commented on src/core_plugins/console/api_server/es_6_0/search.js:
```diff
@@ -1,36 +1,6 @@
-      '{indices}/_search',
-      '_search'
-    ],
-    url_params: {
```
Like the `reindex` api, why remove these params?

spalger commented on src/core_plugins/console/api_server/api.js:
```diff
@@ -1,45 +1,52 @@
+    let copiedDescription = {};
     if (this.endpoints[endpoint]) {
-      throw new Error("endpoint [" + endpoint + "] is already registered");
+      console.warn(`Extending endpoint definition: ${endpoint}`);
```
Looks like this `console.warn` was accidentally left in

spalger commented on src/core_plugins/console/api_server/es_6_0/ingest.js:
```diff
@@ -337,50 +334,21 @@ export default function (api) {
-    ]
-  });
 
-  api.addEndpointDescription('_delete_ingest_pipeline', {
```
What's the reasoning behind deleting these suggestions?

bmcconaghy commented:
I don't believe I have deleted any suggestions.  The entire 5.0 API was deleted and replaced with 6.0.  The generated specs contain all of the completions AFAIK.  Maybe we can chat about this.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2846/)

bmcconaghy commented:
Verified these endpoints aren't deleted:

![image](https://user-images.githubusercontent.com/2479496/39782486-5a558774-52e0-11e8-96c7-a46044cca8a4.png)

We just don't need to override some of them anymore (that's what files in that es_6_0 directory do).

@spalger

spalger commented:
There's a slight misalignment of the toolbar at the right of the active line in console, pretty sure the play button and wrench should be a couple pixels up

![image](https://user-images.githubusercontent.com/1329312/39784879-eec039c0-52ce-11e8-9756-fea3a254dbc4.png)

spalger commented:
Since I just learned about ctrl+space, I'm not sure if this used to work, but ctrl+space and selecting a query string param now replaces the `?`

![2018-05-08 14 46 04](https://user-images.githubusercontent.com/1329312/39784931-1aef3154-52cf-11e8-87f1-6326b226c268.gif)

spalger commented:
There is also some sort of race condition that causes the parameter I have types to only sometimes be selected in the autocomplete list if I just type `?size` at normal typing speed it is actually selected when I'm done about 25% of the time.

![image](https://user-images.githubusercontent.com/1329312/39784986-4d5a0498-52cf-11e8-8127-ab0c3ef04732.png)

spalger commented:
I need to get two versions of console running to make sure these things are actually regressions. I'm sure there are plenty of bugs in the console that i might just be cataloging in this issue.

w33ble commented:
Oddly, when I run this, I get some warning about cursor position from `node_modules/react-ace/node_modules/brace/index.js`, despite Console seemingly relying on the local brace package. Maybe this is a bug in the optimizer? The versions differ though (0.11.0 as opposed to the 0.10.0 that we have in the root package.json file).

![screenshot 2018-05-08 15 47 45](https://user-images.githubusercontent.com/404731/39786815-313d41f0-52d7-11e8-8faa-9c9dd42c4fb5.png)

This is a warning I don't get on master.

bmcconaghy commented on src/core_plugins/console/public/src/sense_editor/mode/input.js:
```diff
@@ -1,27 +1,25 @@
 
-var HighlightRules = require("./input_highlight_rules").InputHighlightRules;
-
-acequire("ace/config").setModuleUrl("sense_editor/mode/worker", require("file-loader!./worker.js"));
```
The file loader actually would not work with the new(er) ace, could not for the life of me figure out why.  I had to follow the pattern of the built in modes and do exports.source='<ALL_THE_SOURCE_AS_A_STRING!>';

spalger replied:
If ace is expecting the module argument to be like `{ source: 'js...' }` then maybe we should use:

```js
const workerModule = {
  source: require('raw-loader!./worker.js')
};

new WorkerClient(['ace', 'sense_editor'], workerModule, 'SenseWorker', 'sense_editor/mode/worker')
```

That way webpack will take care of converting the module to a string.

bmcconaghy replied:
oh yeah that would be nicer

bmcconaghy commented on src/core_plugins/console/public/src/sense_editor/mode/input.js:
```diff
@@ -52,17 +50,14 @@ oop.inherits(Mode, TextMode);
-
   this.createWorker = function (session) {
-    var worker = new WorkerClient(["ace", "sense_editor"], "sense_editor/mode/worker", "SenseWorker");
+    const worker = new WorkerClient(['ace', 'sense_editor'], require('./worker.js'), 'SenseWorker', 'sense_editor/mode/worker');
```
evolution of figuring it out.

bmcconaghy commented:
@w33ble I've seen that warning with other things that use ace editor (index management for example).  Not sure what it's about but seems benign.

The mixing of import and require is due to incomplete refactor to new school import.  In some specific cases it is necessary to use require, and so I'm hesitant to make many changes in that area (tried and saw some weird things).

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2869/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2887/)

spalger commented on x-pack/plugins/logstash/public/sections/pipeline_edit/components/pipeline_edit/pipeline_edit.js:
```diff
@@ -49,25 +46,24 @@ app.directive('pipelineEdit', function ($injector) {
-           * edges of the top/bottom lines to make for a less-crowded
-           * typing experience.
-           */
+-           * This sets the space between the editor's borders and the
```
Pretty sure these `-` shouldn't be here

bmcconaghy replied:
lol yeah

spalger commented on x-pack/plugins/logstash/public/sections/pipeline_edit/components/pipeline_edit/pipeline_edit.js:
```diff
@@ -49,25 +46,24 @@ app.directive('pipelineEdit', function ($injector) {
           );
-
           editor.setReadOnly(this.isReadOnly);
+          editor.getSession().setMode('brace/mode/ruby');
```
Why add this?

bmcconaghy replied:
might be a bad merge

bmcconaghy commented:
@spalger I have addressed all the concerns except for the unnecessary quotes in keys in js objects.  I want to separate that out into its own work effort focused on bringing the code into conformity with our eslint rules and eliminating the custom rules where possible.

spalger commented on src/core_plugins/console/api_server/spec/overrides/indices.create.json:
```diff
@@ -1,8 +1,5 @@
+  "indices.create": {
     "data_autocomplete_rules": {
       "mappings": {
         "__scope_link": "_put_mapping"
```
looks like `_put_mapping` used to be an endpoint defined here: https://github.com/elastic/kibana/pull/18930/files#diff-5a9db413dff420db68b109840db90543L177. Do you think we can check that all of the `__scope_link` references line up with a script or something?

bmcconaghy replied:
good idea

spalger commented on src/core_plugins/console/public/src/autocomplete.js:
```diff
@@ -9,11 +9,11 @@ import { populateContext } from './autocomplete/engine';
 import _ from 'lodash';
 import ace from 'ace';
-import 'ace/ext-language_tools';
+import 'brace/ext/language_tools';
```
IIRC `brace` and `ace` are not the same module. Shouldn't we be using one or the other?

bmcconaghy replied:
This used to work because everything was copied into webpackShims under ace.  brace calls itself ace.  it's all confusing, but what is there works.

spalger commented on src/core_plugins/console/public/src/autocomplete.js:
```diff
@@ -987,6 +987,7 @@ export default function (editor) {
     // change starts here
     var token = session.getTokenAt(pos.row, pos.column);
     this.base = _.clone(pos);
+    this.base.detach = () => {};
```
Do you know if there is some cleanup we should be doing here? Why is a `noop` function sufficient? A comment in the body of the function would be nice.

bmcconaghy replied:
the detach hook was added as part of the new version.  I added this noop to fix a fatal issue.

spalger replied:
I think it's important to understand what the `detach` function on positions normally does to make sure we don't need to do something here. Just because there isn't a fatal issue doesn't mean that this works as it should.

bmcconaghy replied:
detach is defined elsewhere in the code to detach event listeners.  as we are not creating any here, i think it is safe.

spalger commented on src/core_plugins/console/public/src/input.js:
```diff
@@ -1,20 +1,20 @@
-let utils = require('./utils');
-let es = require('./es');
-let history = require('./history');
+require('brace/ext/searchbox');
```
again mixing `ace` and `brace`

bmcconaghy replied:
import ace ... is brace.

spalger replied:
Ah, I see, `import 'ace'` loads the ace webpack shim which loads brace.

spalger commented on src/core_plugins/console/public/src/sense_editor/mode/input.js:
```diff
@@ -1,27 +1,24 @@
-let acequire = require('acequire');
-require('ace');
-require('ace/mode-json');
+import ace from 'ace';
```
We're no longer requiring `ace/mode-json`, is that intentional?

bmcconaghy replied:
We don't seem to need to.  The syntax checking works without it.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2915/)

bmcconaghy commented:
Thank you for the thorough review @spalger, you caught a lot of mistakes.

w33ble commented:
I'll echo that. I tried out a bunch of things, and they all worked just fine.

w33ble commented:
Any chance anyone is able to confirm this closes https://github.com/elastic/kibana/issues/11608?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3047/)

spalger commented on src/core_plugins/console/public/src/sense_editor/mode/worker.js:
```diff
@@ -0,0 +1,2 @@
+exports.id = 'sense_editor/mode/worker';
+exports.src = '(function(t){if(!(void 0!==t.window&&t.document||t.acequire&&t.define)){t.console||(t.console=function(){let t=Array.prototype.slice.call(arguments,0);postMessage({type:\"log\",data:t})},t.console.error=t.console.warn=t.console.log=t.console.trace=t.console),t.window=t,t.ace=t,t.onerror=function(t,e,n,r,i){postMessage({type:\"error\",data:{message:t,data:i.data,file:e,line:n,col:r, ...
```
Did you mean to revert this?

bmcconaghy replied:
yeah.  this was not working as I thought it did when I tested it.  Working on better solution.  Weirdly raw-loader adds use strict to the top of the file and this seems to break things AFAICT.

spalger replied:
That's probably because it was being run through babel, if you use `require('!!raw-loader!./worker.js')` the default loaders for `.js` files will not be applied

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3124/)

elasticmachine commented:
Pinging @elastic/es-ui
